src/bun.js/test/jest.zig

```diff
@@ -135,8 +137,56 @@ pub const TestRunner = struct {
         }
     }
 
-    pub fn hasTestFilter(this: *const TestRunner) bool {
-        return this.filter_regex != null;
+    pub inline fn hasTestFilter(this: *const TestRunner) bool {
+        return this.filter_regex != null or this.hasTestLineFilter();
+    }
+
+    pub inline fn hasTestLineFilter(this: *const TestRunner) bool {
+        return this.line_filters_by_file_id.count() > 0;
+    }
+
+    pub inline fn needsTestLineNumber(this: *const TestRunner) bool {
+        return this.test_options.file_reporter == .junit or this.hasTestLineFilter();
+    }
+
+    pub inline fn shouldSkipBasedOnLineFilter(this: *const TestRunner, line_number: u32, parent: *DescribeScope) bool {
+        return this.hasTestLineFilter() and !this.matchesLineFilter(line_number, parent);
+    }
+
+    pub fn matchesLineFilter(this: *const TestRunner, line_number: u32, parent: *DescribeScope) bool {
+        if (this.line_filters_by_file_id.count() == 0) return true;
+
+        const lines = this.line_filters_by_file_id.get(parent.file_id) orelse return true;
+        for (lines.items) |filter_line| {
+            if (line_number == filter_line) return true;
+
+            var current_parent: ?*DescribeScope = parent;
+            while (current_parent) |p| {
+                if (p.line_number > 0 and p.line_number == filter_line) {
+                    return true;
+                }
+                current_parent = p.parent;
+            }
+        }
+        return false;
+    }
+
+    pub fn addLineFilters(this: *TestRunner, line_filters: bun.StringHashMap(std.ArrayList(u32))) void {
```

Comment:
Can you have a TestLineFilter struct that has the path: []const u8, line: u32 and implements a hash function and an equal function, which is used for  a HashMap context? And then it can expose a Map struct that encapsulates all the logic for TestLineFilter? 

What you have here is nested allocations that would make it easy to have a use-after-resize due to stale pointers. We can instead make this a single hash map used by all of the files in the list this way. 

I'm assuming the way this works will make it exclude files not in the list entirely which means they won't get evaluated and therefore every time the list is non-empty it would be frequently reached.

Reply:
Done, I wonder the perf of this method. But its very few files that are going be in it, so it indeed is better to be safer.

I wonder if its worth implementing range (vitest doesn't have support that), as one easy way would be add many entries for the range.